### ISSUE TITLE
Add support for clojure.spec

### ISSUE THREAD

**Issue by danielcompton** [cursive-ide/cursive#1427]
I know you're looking at this already, but this is a placeholder for people to add their +1's to.

**Comment by nahuel** [comment]
+1

**Comment by seantempesta** [comment]
+1

**Comment by wilkerlucio** [comment]
@cursive-ide I have one suggestion for a feature on this integration. Would be nice if you index the spec information on keywords, so when we cmd + mouse over on a qualified keyword, we could see it's specs, that could be very helpful to understand data as you read the program.

**Comment by nahuel** [comment]
it will be better to use the clojure.spec information to show validation errors and offer smart completions, for example when editing an spec'ed data structure without any function/macro call.

**Comment by mtruyens** [comment]
I hope that it will also be possibly to "resolve" a custom function into either s/def or s/fdef (similar to how you can "resolve" a custom function as a let, defn, def, -> or ->> form). I have defined some macros that take the same inputs as s/def and s/fdef but  facilitate easier testing of input/output values — and it would be cool if those functions would be recognized by Cursive.

**Comment by mtruyens** [comment]
Another very useful feature, I think, would be to specify a custom background and/or foreground color for s/def or s/fdef forms. Right now those forms constitute "visual noise" if you are merely scrolling through a file and are not interested in the actual specs. Having a separate color for those forms would allow to mentally tune out those parts.

**Comment by cursive-ide** [comment]
Originally I thought that it would be simpler to restrict the list of forms which custom forms can be resolved as, but I'm starting to think that it might make sense to just allow anything. Perhaps they could be presented as the current list of common options, with a "Custom..." option at the end which lets you specify anything.

@mtruyens It sounds like what you'd really want for that would be code folding, so that those forms could be folded by default with just enough information to see what you're spec'ing. Does that sound like a useful solution?

**Comment by mtruyens** [comment]
It would indeed be very cool to allow anything for resolvance. For example, I would be very interested in having a macro to write Specs in a Schema format, as this allows for much more compact functions. 

For example, instead of having

```
(defn sample [x y] ...) 
(s/fdef sample :args (s/cat :x integer? :y string?), :ret map?)

```
I could then write

```
(mydefn sample :- map?
[x :- integer?, :y string?]
...)
```

But in order for this to be useful in Cursive, one should be able to resolve "mydefn" as the Schema-resolver.

**Comment by mtruyens** [comment]
As regards the code folding — this would be a very good alternative to the custom background/foreground color, if combined with some shortcut to fold/unfold all the Specs in the current namespace at once.

**Comment by cursive-ide** [comment]
Right, IntelliJ has a lot of commands for folding and unfolding out of the box, see [here](https://www.jetbrains.com/help/idea/2016.2/code-folding.html).

**Comment by Mechrophile** [comment]
Spec seems like an automated refactoring dream come true.  How about the ability to generate a function stub from an fspec

**Comment by gnl** [comment]
@cursive-ide  - just allowing the fully qualified spec keywords to be considered symbols would be a big help, so that after doing (s/def ::something ...) I can view or jump to the definition on every occurrence of ::something as well as use structural/symbol navigation.

Right now I can tell cursive to resolve s/def as def, but it seems as though it doesn't consider the keyword a valid identifier and I'm forced to do normal searches to get to the spec or keep it open in a separate pane.

There's a ton of other cool stuff that could be done with the spec support, but as a first step this seems like the biggest bang for the buck.

**Comment by marco-m** [comment]
Hello @cursive-ide any idea of a timeline for initial/basic support of clojure.spec ? Thanks!

**Comment by gnl** [comment]
@marco-m - the most important thing - view+go to spec definition - already works.

**Comment by eoliphan** [comment]
Also, would be nice have EDN support such that when editing `foo.edn` specs from your project code and/or libs would be applied.  Would really be great for supporting Clojure "Data"SL's   Supporting spec enhancers like expound would be useful as well
